### PR TITLE
Initial support for xy

### PR DESCRIPTION
**Pull request by brucemiller**
As it says. Actually it covers a lot of stuff, and some of it quite well. Alas, there are significant bits that are off, in particular badly positioned. Consequently, it isn't quite ready for prime time.  

See the test `xytests` in the `LaTeXML-tests` repo.

### CODE REVIEW

File: lib/LaTeXML/Package/xylatexml.tex.ltxml [hunk 1]
```diff
@@ -0,0 +1,994 @@
...
+DefMacro('\bhook@@',      '\lx@xy@hook{-1}');                              # "(" tip shifted down
+DefMacro('\aturn@@',      '\lx@xy@turn{1}');                               # quarter circle tip  up
+DefMacro('\bturn@@',      '\lx@xy@turn{-1}');                              # quarter circle tip down
+DefMacro('\solidpoint@',  '\pointlike@{\lx@xy@fill@on\lx@xy@point}\jot');  # filled circle
+DefMacro('\hollowpoint@', '\pointlike@{\lx@xy@fill@off\lx@xy@point}\jot'); # empty circle
+
+# Redefined to position w/o kerns.
+RawTeX(<<'EoTeX');
+%\xydef@\Tip@{\kern2.5pt \vrule height2.5pt depth2.5pt width\z@
+% \Tip@@ \kern2.5pt \egroup
```
Comment (dginev) [inline lib/LaTeXML/Package/xylatexml.tex.ltxml:393]: Generally, could we take out the `%` comment lines out of the RawTeX and into a perl `#` comment?

Saves on runtime, but more importantly it also makes the RawTeX itself easier to review, as I don't have to double-guess typos.

**Review by dginev — COMMENTED** [review]
The code of xylatexml.tex.ltxml looks quite impressive on a surface read. Both in size, and in how well-organized it is.

I'll take the tests for a spin and see if I can learn more while getting my hands dirty.

**Comment by dginev** [comment]
Update for @brucemiller : 

Hi! I have finally created and deployed [an arXiv sandbox of XY use](https://corpora.mathweb.org/corpus/sandbox%5Farxiv%5Fxy/tex%5Fto%5Fhtml), with 9775 articles.

I am now running through it with the latest master branch of latexml, to get a baseline success rate.

May I request that you run a quick `git rebase origin/master` on the branch for this PR, and push it to github? Then we can compare the results knowing that the only changes are the diffs displayed here, for some extra sanity.

Thanks!

**Comment by brucemiller** [comment]
done!

**Comment by brucemiller** [comment]
Are the ar5iv workers (or some of them) still using the older stub xy?

**Comment by dginev** [comment]
... I may need to really consider fully deleting the docker images before I update them, since it looks like some of the existing XY files are still on disk. Will rerun, thank you for spotting that!

**Comment by dginev** [comment]
@brucemiller prefacing with "I am not making this up - people really write this!"

I updated with the latest commit, and the two documents (and a handful more) now convert neatly without Fatals.

And yet, then  I encountered 1912.10009, which still hits the same timeout snag, *but* with an extra layer of indirection. Namely, a setup:

main.tex
```tex
\documentclass{article}
\usepackage{test}
\begin{document}
\end{document}
```

with a custom test.sty reducing to:
```tex
\usepackage[all]{xy}
\input xy
\xyoption{all}
```

We probably ought to add a guard for this case...

**Comment by dginev** [comment]
The "too many errors" are quite gnarly and hard to debug (some articles have >5,000 lines of TeX!), but I found a couple that exhibited the same problem that brought about 100 errors.

It's a rather unhappy one though, as it touches catcode activation scoping.

Here is a somewhat minimal example:
```tex
\documentclass{article}
\usepackage{amssymb}
\makeatletter
\usepackage[all]{xy}

\begin{document}

$$ \xymatrix{ & ^{\mathbb{P}^1}\ar@{-}[ddd]^(.3){\mathfrak{G}[p]}&  &
      ^{\mathbb{P}^1}\ar@{-}[ddd]^(.3){\mathfrak{G}[p]}&^{\mathbb{P}^1}\ar@{-}[ddd]^(.3){\mathfrak{G}[p]} & \\
   & &\dots & & & \\
   \ar@{-}[rrrrr]^(.35){\kappa\otimes\alpha_p}&\bullet & & \bullet&\bullet & ^{\mathbb{P}^1}\\
   & & ^{\zeta}\ar@/^/@{..>}[ru] \ar@/^/@{..>}[rru] \ar@/_/@{..>}[lu] & & &  \\
   } $$

\end{document}
```

Interestingly, removing the `\makeatletter` gets the conversion to complete without any issues at all, and the diagram is _starting to_ look right.

pdflatex encounters no issues with or without `\makeatletter`.

The original article is 1703.05720.

**Comment by dginev** [comment]
I am inspecting the HTML for a surprisingly successful job, where XY and Tikz were bravely combined by an arXiv author. The example in question is in the following details tag:

<details>

```tex
\documentclass{article}
% xypic
\usepackage[all]{xy}
\SelectTips{cm}{}
%%% tikz
\usepackage{tikz}
\usetikzlibrary{decorations.markings}
\usetikzlibrary{decorations.pathreplacing}
\usetikzlibrary{arrows,shapes,positioning}
\tikzstyle directed=[postaction={decorate,decoration={markings,
    mark=at position #1 with {\arrow{>}}}}]
\tikzstyle rdirected=[postaction={decorate,decoration={markings,
    mark=at position #1 with {\arrow{<}}}}]
\begin{document}
\begin{equation}\label{eqn-decmigr}
\xy
(0,0)*{
\begin{tikzpicture} [scale=.5,fill opacity=0.2]
	%shading
	\path[fill=green] (2.25,3) to (.75,3) to (.75,0) to (2.25,0);
	\path[fill=green] (.75,3) to [out=225,in=0] (-.5,2.5) to (-.5,-.5) to [out=0,in=225] (.75,0);
	\path[fill=green] (.75,3) to [out=135,in=0] (-1,3.5) to (-1,.5) to [out=0,in=135] (.75,0);
	%bottom web
	\draw [very thick,directed=.55] (2.25,0) to (.75,0);
	\draw [very thick,directed=.55] (.75,0) to [out=135,in=0] (-1,.5);
	\draw [very thick,directed=.55] (.75,0) to [out=225,in=0] (-.5,-.5);
	%seam
	\draw[very thick, red, directed=.55] (.75,0) to (.75,3);
	%vertical edges
	\draw [very thick] (2.25,3) to (2.25,0);
	\draw [very thick] (-1,3.5) to (-1,.5);
	\draw [very thick] (-.5,2.5) to (-.5,-.5);
	%top web
	\draw [very thick,directed=.55] (2.25,3) to (.75,3);
	\draw [very thick,directed=.55] (.75,3) to [out=135,in=0] (-1,3.5);
	\draw [very thick,directed=.55] (.75,3) to [out=225,in=0] (-.5,2.5);
	%labels
	\node [black, opacity=1]  at (1.5,2.75) {\tiny{$_{a+b}$}};
	\node[black, opacity=1] at (-.75,3.25) {\tiny{$b$}};
	\node[black, opacity=1] at (-.25,2.25) {\tiny{$a$}};
	\node[opacity=1] at (1.5,1.5) {\small$\pi_\gamma$};
\end{tikzpicture}
};
\endxy
\quad = \sum_{\alpha,\beta} c_{\alpha,\beta}^\gamma \quad
\xy
(0,0)*{
\begin{tikzpicture} [scale=.5,fill opacity=0.2]
	%shading
	\path[fill=green] (2.25,3) to (.75,3) to (.75,0) to (2.25,0);
	\path[fill=green] (.75,3) to [out=225,in=0] (-.5,2.5) to (-.5,-.5) to [out=0,in=225] (.75,0);
	\path[fill=green] (.75,3) to [out=135,in=0] (-1,3.5) to (-1,.5) to [out=0,in=135] (.75,0);
	%bottom web
	\draw [very thick,directed=.55] (2.25,0) to (.75,0);
	\draw [very thick,directed=.55] (.75,0) to [out=135,in=0] (-1,.5);
	\draw [very thick,directed=.55] (.75,0) to [out=225,in=0] (-.5,-.5);
	%seam
	\draw[very thick, red, directed=.55] (.75,0) to (.75,3);
	%vertical edges
	\draw [very thick] (2.25,3) to (2.25,0);
	\draw [very thick] (-1,3.5) to (-1,.5);
	\draw [very thick] (-.5,2.5) to (-.5,-.5);
	%top web
	\draw [very thick,directed=.55] (2.25,3) to (.75,3);
	\draw [very thick,directed=.55] (.75,3) to [out=135,in=0] (-1,3.5);
	\draw [very thick,directed=.55] (.75,3) to [out=225,in=0] (-.5,2.5);
	%labels
	\node[black, opacity=1]  at (1.5,2.75) {\tiny{$_{a+b}$}};
	\node[black, opacity=1] at (-.75,3.25) {\tiny{$b$}};
	\node[black, opacity=1] at (-.25,2.25) {\tiny{$a$}};
	\node[opacity=1] at (.125,1.25) {\small$\pi_\alpha$};
	\node[opacity=1] at (-.25,3) {\small$\pi_\beta$};
\end{tikzpicture}
};
\endxy
\end{equation}
\end{document}
```

</details>

1. The first piece that struck me is that the open `<math>` tag in the HTML had a 600 line long `alttext` attribute, almost entirely filled with drawing artefacts from XY and PGF. Would this be a good time to suggest that we change the attribute reversions for the drawing packages, so that they deposit a smaller "blackboxy" description of the graphic? 
    - The level of verbosity would harm use in accessibility, given it is TeX-near and not reader-near. I can only imagine it useful in "reproducing" the graphic by running pdflatex (or latexml) over the alttext to regenerate the HTML - but I am unsure we want to be encouraging such use of alttext? (I am also unsure this holds in practice).

2. In the HTML we have another case where the SVG renders as "almost right", but there are various positioning subtleties that have gone astray.

Hopefully a useful example!

**Comment by dginev** [comment]
Attaching one more isolated example where latexml encounters 100 errors (Fatal), and pdflatex is happy. Source document is 1006.4120.

<details>

```tex
\documentclass{article}
\usepackage[all]{xy}
\usepackage{amsmath,amssymb,latexsym}

\begin{document}

\begin{equation}
\xymatrix{
    \boxed{\mathcal{V}_{0,0}} \ar@<1ex>[r]^{f^{+}} \ar@<1ex>[d]^{b^{+}} & \boxed{\mathcal{V}_{0,1}} \ar@<1ex>[r]^{f^{+}} \ar@<1ex>[l]^{f^{-}} \ar@<1ex>[d]^{b^{+}} & \ldots \ar@<1ex>[r]^{f^{+}} \ar@<1ex>[l]^{f^{-}} &  \boxed{\mathcal{V}_{0,n}} \ar@<1ex>[r]^{f^{+}} \ar@<1ex>[l]^{f^{-}} \ar@<1ex>[d]^{b^{+}} & \ldots \ar@<1ex>[l]^{f^{-}} & \ldots \ar@<1ex>[r]^{f^{+}} & \boxed{\mathcal{V}_{0,p-1}} \ar@<1ex>[r]^{f^{+}} \ar@<1ex>[l]^{f^{-}} \ar@<1ex>[d]^{b^{+}} & \boxed{\mathcal{V}_{0,p}} \ar@<1ex>[l]^{f^{-}} \ar@<1ex>[d]^{b^{+}} \\
    \boxed{\mathcal{V}_{1,0}} \ar@<1ex>[r]^{f^{+}} \ar@<1ex>[d]^{b^{+}} \ar@<1ex>[u]^{b^{-}} & \boxed{\mathcal{V}_{1,1}} \ar@<1ex>[r]^{f^{+}} \ar@<1ex>[l]^{f^{-}} \ar@<1ex>[d]^{b^{+}} \ar@<1ex>[u]^{b^{-}} & \ldots \ar@<1ex>[r]^{f^{+}} \ar@<1ex>[l]^{f^{-}} &  \boxed{\mathcal{V}_{1,n}} \ar@<1ex>[r]^{f^{+}} \ar@<1ex>[l]^{f^{-}} \ar@<1ex>[d]^{b^{+}} \ar@<1ex>[u]^{b^{-}} & \ldots \ar@<1ex>[l]^{f^{-}} &  \ldots \ar@<1ex>[r]^{f^{+}} & \boxed{\mathcal{V}_{1,p-1}} \ar@<1ex>[r]^{f^{+}} \ar@<1ex>[l]^{f^{-}} \ar@<1ex>[d]^{b^{+}} \ar@<1ex>[u]^{b^{-}} & \boxed{\mathcal{V}_{1,p}} \ar@<1ex>[d]^{b^{+}} \ar@<1ex>[u]^{b^{-}} \ar@<1ex>[l]^{f^{-}} \\
    \vdots \ar@<1ex>[d]^{b^{+}} \ar@<1ex>[u]^{b^{-}} & \vdots \ar@<1ex>[d]^{b^{+}} \ar@<1ex>[u]^{b^{-}} & \ldots  & \vdots \ar@<1ex>[d]^{b^{+}} \ar@<1ex>[u]^{b^{-}}  & \ldots \ar@<1ex>[d]^{b^{+}}  & \ldots  & \vdots \ar@<1ex>[u]^{b^{-}}  & \vdots \ar@<1ex>[d]^{b^{+}} \ar@<1ex>[u]^{b^{-}}    \\
    \boxed{\mathcal{V}_{m,0}} \ar@<1ex>[r]^{f^{+}} \ar@<1ex>[d]^{b^{+}} \ar@<1ex>[u]^{b^{-}}  & \boxed{\mathcal{V}_{m,1}} \ar@<1ex>[r]^{f^{+}} \ar@<1ex>[l]^{f^{-}} \ar@<1ex>[d]^{b^{+}} \ar@<1ex>[u]^{b^{-}} & \ldots \ar@<1ex>[r]^{f^{+}} \ar@<1ex>[l]^{f^{-}} & \boxed{\mathcal{V}_{m,n}} \ar@<1ex>[r]^{f^{+}} \ar@<1ex>[l]^{f^{-}} \ar@<1ex>[d]^{b^{+}} \ar@<1ex>[u]^{b^{-}} & \boxed{\mathcal{V}_{m,n+1}} \ar@<1ex>[r]^{f^{+}} \ar@<1ex>[l]^{f^{-}} \ar@<1ex>[d]^{b^{+}} \ar@<1ex>[u]^{b^{-}} & \ldots \ar@<1ex>[l]^{f^{-}}  & \ldots \ar@<1ex>[r]^{f^{+}} & \boxed{\mathcal{V}_{m,p}} \ar@<1ex>[l]^{f^{-}} \ar@<1ex>[d]^{b^{+}} \ar@<1ex>[u]^{b^{-}} \\
    \vdots \ar@<1ex>[u]^{b^{-}} & \vdots \ar@<1ex>[u]^{b^{-}} &  \ldots \ar@<1ex>[r]^{f^{+}}  & \boxed{\mathcal{V}_{m+1,n}} \ar@<1ex>[r]^{f^{+}} \ar@<1ex>[l]^{f^{-}} \ar@<1ex>[d]^{b^{+}} \ar@<1ex>[u]^{b^{-}} & \ldots \ar@<1ex>[l]^{f^{-}} \ar@<1ex>[u]^{b^{-}} &  \ldots  & \vdots  & \vdots \ar@<1ex>[u]^{b^{-}}    \\
    \vdots & \vdots & \ldots & \vdots & \ldots & \ldots  & \vdots & \vdots
}
\end{equation}

\end{document}
```

</details>

**Comment by brucemiller** [comment]
The same underlying error, I think; it's using ams macros that have been defined as Constructors which is (somehow) interfering with the expansion of the `\halign`, causing stray conditionals (or timing).

**Comment by brucemiller** [comment]
There still seem to be that perl warning (redefined `acos` & similar). Many fewer but still 13%; maybe some workers didn't get updated?

**Comment by brucemiller** [comment]
OK, the above commits should fix the problems reported to this point, and adjust the positioning as well. Perhaps a re-run would be good here to clarifiy the reports?

**Comment by dginev** [comment]
Fantastic! Congratulation on implementing a very substantial piece of the XY functionality entirely on your own!

My first recollection of attempting some version of a XY binding was in the arXMLiv project and dates back to the year 2007. It is very, very impressive to see a functional binding land in the latexml master today.

**Pull request merged**